code/mob/living/critter/ai/dog.dm

```diff
@@ -0,0 +1,107 @@
+/datum/aiHolder/dog
+	New()
+		..()
+		default_task = get_instance(/datum/aiTask/prioritizer/critter/dog, list(src))
+
+/datum/aiTask/prioritizer/critter/dog/New()
+	..()
+	transition_tasks += holder.get_instance(/datum/aiTask/timed/wander, list(holder, src))
+
+// Go fetch!
+/datum/aiTask/sequence/goalbased/critter/dog/fetch
+	name = "fetching"
+	weight = 10
+	max_dist = 7
+	ai_turbo = TRUE
+
+/datum/aiTask/sequence/goalbased/critter/dog/fetch/New(parentHolder, transTask)
+	..(parentHolder, transTask)
+	add_task(holder.get_instance(/datum/aiTask/succeedable/critter/dog/fetch, list(holder)))
+
+/datum/aiTask/sequence/goalbased/critter/dog/fetch/get_targets()
+	var/mob/living/critter/small_animal/dog/the_dog = holder.owner
+	return list(the_dog.fetch_item)
+
+//Fetch subtask, pick up the item
+/datum/aiTask/succeedable/critter/dog/fetch
+	name = "fetch subtask"
+	var/is_complete = FALSE
+
+/datum/aiTask/succeedable/critter/dog/fetch/failed()
+	var/mob/living/critter/C = holder.owner
+	var/obj/item/I = holder.target
+	if(!C || !I || BOUNDS_DIST(I, C) > 0 || !istype(I.loc, /turf)) //the tasks fails and is re-evaluated if the target is not in range
+		return TRUE
+
+/datum/aiTask/succeedable/critter/dog/fetch/succeeded()
+	return is_complete
+
+/datum/aiTask/succeedable/critter/dog/fetch/on_tick()
+	if(!is_complete)
+		holder.stop_move()
+		var/mob/living/critter/C = holder.owner
+		var/obj/item/I = holder.target
+		if(C && I && BOUNDS_DIST(C, I) == 0 && istype(I.loc, /turf))
+			C.set_dir(get_dir(C, I))
+			if(C.set_hand(1))
+				C.drop_item()
+				C.hand_attack(I)
+				if (prob(20))
+					holder.owner.visible_message("<span class='notice'>[holder.owner] begins to chew on [holder.target]!</span>")
+				else
+					holder.owner.visible_message("<span class='notice'>[holder.owner] picks up [holder.target]!</span>")
+					holder.priority_tasks += holder.get_instance(/datum/aiTask/sequence/goalbased/critter/dog/fetch_back, list(holder, holder.default_task))
+			is_complete = TRUE
+
+/datum/aiTask/succeedable/critter/dog/fetch/on_reset()
+	is_complete = FALSE
+
+// C'mere boy!
+/datum/aiTask/sequence/goalbased/critter/dog/fetch_back
+	name = "fetching back"
+	weight = 10
+	max_dist = 7
+	ai_turbo = TRUE
+
+/datum/aiTask/sequence/goalbased/critter/dog/fetch_back/New(parentHolder, transTask)
+	..(parentHolder, transTask)
+	add_task(holder.get_instance(/datum/aiTask/succeedable/critter/dog/fetch_back, list(holder)))
+
+/datum/aiTask/sequence/goalbased/critter/dog/fetch_back/get_targets()
+	var/mob/living/critter/small_animal/dog/the_dog = holder.owner
+	var/mob/living/playmate = the_dog.fetch_playmate
+	if (playmate && get_dist(the_dog, playmate) < max_dist)
+		the_dog.visible_message("<span class='notice'>[the_dog] begins happily running towards [playmate] with their picked up item, wagging their tail furiously!</span>")
```

Comment:
Do we still have a valid reference to target object here?  Wonder if it should be more explicit about the object, at least for small objects.   Tiny objects being obfuscated is kind of interesting.

Reply:
I think that the reference should still be valid? 
When you say the object being more explicit, you mean that it should say "running towards X with [the item name]"?